### PR TITLE
refactor findScalableResourceProviderIDs in clusterapi

### LINKED ISSUE #7976: clusterapi provider can fail to decrease target size when machines have a provider id but fail to become a node
**Which component are you using?**:

/area cluster-autoscaler

**What version of the component are you using?**:

Component version: master

**What k8s version are you using (`kubectl version`)?**: tested with 1.29, 1.30, 1.31

**What environment is this in?**:

clusterapi

**What did you expect to happen?**:

machines that have a provider ID but have been stopped by the cloud provider should not prevent the autoscaler from decreasing the target size of a node group.

**What happened instead?**:

autoscaler fails to reduce the node group, with the following error in the logs:

```
static_autoscaler.go:448] Failed to fix node group sizes: failed to decrease MachineSet/<redacted>: node group <redacted>: attempt to delete existing nodes currentReplicas:10 delta:-1 existingNodes: 10
```

**How to reproduce it (as minimally and precisely as possible)**:

this is difficult to reproduce but will happen in cases where spot instances are used, with provider IDs given, and then stopped by the infrastructure before becoming a node.

**Anything else we need to know?**:

this is related to #7928

### PR DESCRIPTION
**Pull request by elmiko**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

this change refactors the function so that it each distinct machine state can be filtered more easily. the unit tests have been supplemented, but not changed to ensure that the functionality continues to work as expected. these changes are to help better detect edge cases where machines can be transiting through pending phase and might be removed by the autoscaler.

#### Which issue(s) this PR fixes:

Fixes #7976 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by elmiko** [comment]
the failing test is trying to do something that we don't necessarily want. it's removing the node ref from the machine object to force a lookup of the provider id from the machine instead of the node. but, we are changing how the nodes are listed from the provider to ensure that a node without a noderef, but with a provider id, is properly identified as pending.

 i am going to improve these tests to ensure we are getting all the possible outcomes.

**Comment by elmiko** [comment]
after reviewing the [failing test](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go#L1381-L1430), i think we can safely remove this as it is testing functionality that does not comport with the current understanding of how the `Nodes` function should be working.

the test relies on the idea that a Machine object with no `.status.nodeRef.name` will be able to be found by examining the node object and then finding the provider id. it references a function called `machineSetNodeNames` in the comments; this function does not exist. we have changed the functionality of the `Nodes` function to consider Machines without a node ref to be in pending state. due to this change, the provider id for the Machine is returned as normalized with the pending label prefixed. i have changed the test to not remove the noderef as the logic would not allow a machine with no noderef to pass beyond the pending guard.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go [hunk 1]
```diff
@@ -1335,19 +1335,6 @@ func TestControllerMachineSetNodeNamesUsingProviderID(t *testing.T) {
 	controller, stop := mustCreateTestController(t, testConfig)
 	defer stop()
 
-	// Remove Status.NodeRef.Name on all the machines. We want to
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:1338]: Is the idea here that the business logic has now been updated to properly prioritize providerID over Name, and so we don't need to babysit the data input like this anymore?

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 2]
```diff
@@ -648,53 +649,69 @@ func (c *machineController) findScalableResourceProviderIDs(scalableResource *un
...
 		}
 
-		// Next we check for the provider ID. Machines with a provider ID can fall into one of a few
-		// categories: creating, running, deleting, and sometimes failed (but we filtered those earlier).
-		// Depending on which category the machine is in, it might have a deleting or pending guard
-		// added to it's provider ID so that we can later properly filter machines.
+		// Deleting Machines
+		// Machines that are in deleting state should be identified so that in scenarios where the core
+		// autoscaler would like to adjust the size of a node group, we can give a proper count and
+		// be able to filter machines in that state, regardless of whether they become a node or not.
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:665]: I'm not sure a "Deleting machine" would ever become a node, would

> regardless of whether or they are still active nodes in the cluster

be better than

> regardless of whether they become a node or not.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 3]
```diff
@@ -648,53 +649,69 @@ func (c *machineController) findScalableResourceProviderIDs(scalableResource *un
...
-			return nil, err
-		}
-
-		if !found {
-			klog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.GetName())
-			providerIDs = append(providerIDs, createPendingMachineProviderID(machine.GetNamespace(), machine.GetName()))
-			continue
-		}
-
+		// Begin checking to determine if the node reference is valid
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:714]: does the `found` response from `unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "kind")` (L715) need to be taken into consideration before we call `unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "name")` (L725)?

I see we're considering it in L720 (`if found && nodeRefKind != "Node" {`), but is that the only place we want to evaluate it?

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go (hunk 2 above)
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:665]: yes, will update.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go (hunk 3 above)
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:714]: yes, i think we just need to check it that one time. i left this part of the logic unchanged as it seemed to be correct.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go (hunk 1 above)
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:1338]: correct, in the previous iteration, there was a clause to check the provider id and then do different things based on having a provider id and other stuff. since i refactored to be more distinct about failed, deleting, and pending machines, this deep testing of the conditional is not needed.

Comment (jackfrancis) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:1338]: ack, this is a significant improvement in edge case detection, stuff like this in UT is essentially "let's get rid of any possible edge cases because that might be harder to verify" ;)

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go [hunk 4]
```diff
@@ -648,6 +663,23 @@ func TestNodeGroupDecreaseTargetSize(t *testing.T) {
...
+			description:                         "A node group with 4 replicas with one pending machine that has a provider ID should decrease by 1",
+			initial:                             4,
+			targetSizeIncrement:                 0,
+			expected:                            3,
+			delta:                               -1,
+			includePendingMachine:               true,
+			includePendingMachineWithProviderID: true,
+		},
+		{
+			description:                  "A node group with target size 4 but only 3 existing instances without provider IDs should decrease by 1",
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:676]: let's s/should decrease by 1/should not scale out/

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go [hunk 5]
```diff
@@ -406,16 +406,18 @@ func TestNodeGroupIncreaseSize(t *testing.T) {
...
+		delta                               int
+		initial                             int32
+		targetSizeIncrement                 int32
+		expected                            int32
+		expectedError                       bool
+		includeDeletingMachine              bool
+		includeFailedMachine                bool
+		includeFailedMachineWithProviderID  bool
+		includePendingMachine               bool
+		includePendingMachineWithProviderID bool
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:419]: just a note that these UT scenarios could probably be improved in the future by allowing a more flexible count of machines with particular states (right now we only organize our scenarios so they have one of each type of [not sure about the combinatorics] failed/pending/no providerID)

**Comment by elmiko** [comment]
updated with doc string suggestion

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go (hunk 5 above)
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:419]: yes, what i want to do as a next step is to change our unit test infrastructure to have a more fluent style builder for the capi objects. i think that will make the tests easier to understand. this is a middle ground.

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:419]: as to the combinatorics, i have tried to make it so that we can test combinations of machines. i would like to add that during the next step.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go (hunk 4 above)
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go]: updated
